File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/heuristic/assignment/classical/cluster/CoefficientPropagation.java
package cujae.inf.ic.om.heuristic.assignment.classical.cluster;

import java.util.ArrayList;

import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Problem;
import cujae.inf.ic.om.problem.output.solution.Cluster;
import cujae.inf.ic.om.problem.output.solution.Solution;
import cujae.inf.ic.om.service.OSRMService;

import cujae.inf.ic.om.heuristic.assignment.classical.ByNotUrgency;
import cujae.inf.ic.om.matrix.NumericMatrix;
import cujae.inf.ic.om.matrix.RowCol;

public class CoefficientPropagation extends ByNotUrgency {

	public static double degradationCoefficient = 0.5;

	private Solution solution = new Solution();	
	
	private ArrayList<Cluster> listClusters;
	private ArrayList<Customer> listCustomersToAssign;
	private NumericMatrix costMatrix;
	private ArrayList<Double> listCoefficients;
	private ArrayList<ArrayList<Double>> listScaledDistances;
	private NumericMatrix scaledMatrix; 

	public CoefficientPropagation() {
		super();
	}

	@Override
	public Solution toClustering() {
	
		if(degradationCoefficient > 1 || degradationCoefficient < 0)
			degradationCoefficient = 0.5;
		
		initialize();
		assign();
		return finish();
	}	
		
	@Override
	public void initialize() {
		listClusters = initializeClusters();
		listCustomersToAssign = new ArrayList<Customer>(Problem.getProblem().getCustomers());
		costMatrix = initializeCostMatrix(Problem.getProblem().getCustomers(), Problem.getProblem().getDepots(), distanceType);
		listCoefficients = initializeCoefficients();
		listScaledDistances = new ArrayList<ArrayList<Double>>(fillListScaledDistances());
		scaledMatrix = initializeScaledMatrix(listScaledDistances); 
	}

	@Override
	public void assign() {
		int posCustomer = -1;
		int idCustomer = -1;
		double requestCustomer = 0.0;
		
		int posDepot = -1;
		int idDepot = -1;
		double capacityDepot = 0.0;
		
		int posCluster = -1;
		double requestCluster = 0.0;
		
		int posElement = -1;
		RowCol rcBestAll = new RowCol();
		
		int totalItems = listCustomersToAssign.size();
		int totalClusters = Problem.getProblem().getDepots().size();
		
		while((!listCustomersToAssign.isEmpty()) && (!listClusters.isEmpty()) && (!scaledMatrix.fullMatrix(Double.POSITIVE_INFINITY)))
		{
			rcBestAll = scaledMatrix.indexLowerValue();
			
			posCustomer = rcBestAll.getCol(); // puede devolver un deposito?
			idCustomer = Problem.getProblem().getListIDCustomers().get(posCustomer);
			requestCustomer = Problem.getProblem().getRequestByIDCustomer(idCustomer);

			posElement = rcBestAll.getRow();

			if(posElement >= totalItems)	
			{
				idDepot =  Problem.getProblem().getListIDDepots().get((posElement - totalItems)).intValue(); 
				
				posCluster = findCluster(idDepot, listClusters);
				posDepot = posElement;
			}
			else
			{
				posCluster = getPosCluster(posElement, listClusters);
				
				idDepot = listClusters.get(posCluster).getIDCluster();
				posDepot = Problem.getProblem().getPosElement(idDepot);
			}
			
			if(posCluster != -1)
			{
				requestCluster = listClusters.get(posCluster).getRequestCluster();
				capacityDepot = Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepotByIDDepot(idDepot));

				if(capacityDepot >= (requestCluster + requestCustomer))
				{
					requestCluster += requestCustomer;
				
					listClusters.get(posCluster).setRequestCluster(requestCluster);
					listClusters.get(posCluster).getItemsOfCluster().add(idCustomer);

					calculateAttractionCoefficient(posCustomer, posElement, listCoefficients);
					scaledMatrix.fillValue(0, posCustomer, (totalItems + totalClusters - 1), posCustomer, Double.POSITIVE_INFINITY);

					listCustomersToAssign.remove(Problem.getProblem().findPosCustomer(listCustomersToAssign, idCustomer));
					updateScaledMatrix(listCustomersToAssign, posCustomer, scaledMatrix, listCoefficients);
				
					if(isFullDepot(listCustomersToAssign, requestCluster, capacityDepot))
					{
						scaledMatrix.fillValue((posDepot), 0, (posDepot), (totalItems + totalClusters - 1), Double.POSITIVE_INFINITY);
						
						for(int i = 0; i < listClusters.get(posCluster).getItemsOfCluster().size(); i++)
						{
							int posElement1 = Problem.getProblem().getPosElement(listClusters.get(posCluster).getItemsOfCluster().get(i));
							scaledMatrix.fillValue(posElement1, 0, posElement1, (totalItems + totalClusters - 1), Double.POSITIVE_INFINITY);	
						}
						
						if(!(listClusters.get(posCluster).getItemsOfCluster().isEmpty()))
							solution.getClusters().add(listClusters.remove(posCluster));
						else
							listClusters.remove(posCluster);
					}
				}
				else
				{	if(scaledMatrix.fullMatrix(0, posCustomer, totalItems + totalClusters - 1, posCustomer, Double.POSITIVE_INFINITY))
					{
						solution.getUnassignedItems().add(idCustomer);
						listCustomersToAssign.remove(Problem.getProblem().findPosCustomer(listCustomersToAssign, idCustomer));
					}
				
					scaledMatrix.setItem(posElement, posCustomer, Double.POSITIVE_INFINITY);
				}
			}
		}
	}
	
	@Override
	public Solution finish() {
		if(!listCustomersToAssign.isEmpty())					
			for(int j = 0; j < listCustomersToAssign.size(); j++)	
				solution.getUnassignedItems().add(listCustomersToAssign.get(j).getIDCustomer());

		if(!listClusters.isEmpty())
			for(int k = 0; k < listClusters.size(); k++)
				if(!(listClusters.get(k).getItemsOfCluster().isEmpty()))
					solution.getClusters().add(listClusters.get(k));
		
		OSRMService.clearDistanceCache();
		
		return solution;
	}

	/*Este método calcula el coeficiente de atracción dados, el identificador del cliente asignado, y el cliente por el que se asignó*/
	private void calculateAttractionCoefficient(int posCustomer, int posElement, ArrayList<Double> coefficients){
		double currentAttCoeff = 1.0;
		double newAttCoeff = 1.0;

		if(posElement < coefficients.size())
		{
			currentAttCoeff = coefficients.get(posElement);
			newAttCoeff = Math.min(1, (currentAttCoeff + (currentAttCoeff * degradationCoefficient)));
		}
		else
			newAttCoeff = 1;
		
		coefficients.set(posCustomer, newAttCoeff);
	}
	
	/*Este método inicializa la lista de coeficientes de atracción de los clientes a 0. La lista que devuelve es paralela a la lista de clientes*/
	private ArrayList<Double> initializeCoefficients(){
		ArrayList<Double> coefficients = new ArrayList<Double>();

		int totalItems = Problem.getProblem().getTotalCustomers();

		for(int i = 0; i < totalItems; i++)
			coefficients.add(0.0);

		return coefficients;
	}
	
	/*Este método se encarga de llenar una lista de listas con las distancias escaladas, para luego crear la NumericMatrix con estos datos*/
	private ArrayList<ArrayList<Double>> fillListScaledDistances(){
		ArrayList<ArrayList<Double>> scaledDistances = new ArrayList<ArrayList<Double>>();

		ArrayList<Integer> listIDCustomers = new ArrayList<Integer>(Problem.getProblem().getListIDCustomers());
		ArrayList<Integer> listIDDepots = new ArrayList<Integer>(Problem.getProblem().getListIDDepots());

		int totalItems = listIDCustomers.size();
		int totalClusters = listIDDepots.size();

		for(int i = 0; i < totalItems; i++)
		{	  
			ArrayList<Double> listDistances = new ArrayList<Double>();

			for(int j = 0; j < (totalItems + totalClusters); j++)
				listDistances.add(Double.POSITIVE_INFINITY);

			scaledDistances.add(listDistances);
		}
		int posCustomerMatrix;
		int posDepotMatrix;

		for(int i = 0; i < totalClusters; i++)
		{
			ArrayList<Double> listDistances = new ArrayList<Double>();
			posDepotMatrix = Problem.getProblem().getPosElement(listIDDepots.get(i));

			for(int j = 0; j < totalItems; j++)
			{
				posCustomerMatrix = Problem.getProblem().getPosElement(listIDCustomers.get(j));
				listDistances.add(costMatrix.getItem(posCustomerMatrix, posDepotMatrix));//ahora viendolo de nuevo, no estoy segura si es asi o fila posDepotMatrix, y la columna la del cliente
			}
			for(int k = totalItems; k < (totalItems + totalClusters); k++)
				listDistances.add(Double.POSITIVE_INFINITY);

			scaledDistances.add(listDistances);
		}

		return scaledDistances;
	}
	
	/* Método encargado de llenar la matriz de costo usando listas de distancias*/
	private NumericMatrix initializeScaledMatrix(ArrayList<ArrayList<Double>> scaledDistances) {
		int size = scaledDistances.size(); 
		NumericMatrix scaledMatrix = new NumericMatrix(size, size);

		int row = -1;
		int col = -1;
		double costInDistance = 0.0;

		for(int i = 0; i < scaledDistances.size(); i++)
		{
			row = i;

			for(int j = 0; j < scaledDistances.get(i).size(); j++)
			{
				col = j;	
				costInDistance = scaledDistances.get(i).get(j);
				scaledMatrix.setItem(row, col, costInDistance);	
			}
		}
		return scaledMatrix;
	}
	
	/*Este método se encarga de actualizar la matriz de distancias escaladas, dado el nuevo cliente asignado, y los clientes que quedan por asignar*/
	private void updateScaledMatrix(ArrayList<Customer> customersToAssign, int posCustomer, NumericMatrix scaledMatrix, ArrayList<Double> coefficients) {
		int posNewCustomer = -1;
		double scaledDistance = 0.0;
		double distance = 0.0;	

		for(int i = 0; i < customersToAssign.size(); i++)
		{
			posNewCustomer = Problem.getProblem().getPosElement(customersToAssign.get(i).getIDCustomer());
			distance = costMatrix.getItem(posNewCustomer, posCustomer);
			scaledDistance = distance * coefficients.get(posCustomer);

			scaledMatrix.setItem(posCustomer, posNewCustomer, scaledDistance);
		}
	}
	
	/*Este método devuelve la posición del cluster (en la lista de clusters) al que debe asignarse el cliente que se está analizando, dado el identificador del elemento por el que se va a asignar dicho cliente*/
	private int getPosCluster(int posCustomer, ArrayList<Cluster> clusters){
		int posCluster = -1;
		int idCustomer = Problem.getProblem().getListIDCustomers().get(posCustomer);
		int i = 0;
		int j;
		boolean found = false;

		while((i < clusters.size()) && (!found))
		{
			j = 0;
			if(!clusters.get(i).getItemsOfCluster().isEmpty())
			{
				while((j < clusters.get(i).getItemsOfCluster().size()) && (!found))
				{
					if(clusters.get(i).getItemsOfCluster().get(j) == idCustomer)
					{
						posCluster = i;
						found = true;
					}
					else
						j++;
				}
			}
			i++;
		}
		return posCluster;
	}
}

File: BHAVRP_Python/cujae/inf/citi/om/assignment/classical/cluster/coefficient_propagation.py
import numpy as np
from typing import List
from ..by_not_urgency import ByNotUrgency
from ....service.osrm_service import OSRMService
from ....problem.input.problem import Problem
from ....problem.input.customer import Customer
from ....problem.output.solution import Solution
from ....problem.output.cluster import Cluster

class CoefficientPropagation(ByNotUrgency):
    degradation_coefficient = 0.5
    
    def __init__(self):
        super().__init__()
        self.solution = Solution()
        self.list_clusters: List[Cluster]
        self.list_customers_to_assign: List[Customer]
        self.list_coefficients: List[float]
        self.list_scaled_distances: List[List[float]]
        self.scaled_matrix: np.ndarray

    def to_clustering(self):
        if not (0 <= self.degradation_coefficient <= 1):
            self.degradation_coefficient = 0.5
        
        self.initialize()
        self.assign()
        return self.finish()  
    
    def initialize(self):
        self.list_clusters = self.initialize_clusters()
        self.list_customers_to_assign = list(Problem.get_problem().get_customers())
        self.list_coefficients = self.initialize_coefficients()
        self.list_scaled_distances = self.fill_list_scaled_distances()
        self.scaled_matrix = self.initialize_scaled_matrix()

    def assign(self):
        total_items = len(self.list_customers_to_assign)
        cost_matrix: np.ndarray = self.initialize_cost_matrix(self.list_customers_to_assign, Problem.get_problem().get_depots(), self.distance_type)
        
        while(any(self.list_customers_to_assign) and self.list_clusters):
            min_value = np.min(cost_matrix) 
            row, col = np.where(cost_matrix == min_value)
            row, col = row[0], col[0]
            
            id_customer = Problem.get_problem().get_list_id_customers()[col]
            request_customer = Problem.get_problem().get_request_by_id_customer(id_customer)
            
            if (row + total_items) >= total_items:
                id_depot = Problem.get_problem().get_list_id_depots()[row]
                pos_cluster = self.find_cluster(id_depot, self.list_clusters)
                pos_depot = row
            else:   
                pos_cluster = self.get_pos_cluster(row, self.list_clusters)
                id_depot = self.list_clusters[pos_cluster].get_id_cluster()
                pos_depot = Problem.get_problem().get_pos_element(id_depot)
            
            if pos_cluster != -1:
                cluster = self.list_clusters[pos_cluster]
                request_cluster = cluster.get_request_cluster()
                capacity_depot = Problem.get_problem().get_total_capacity_by_depot(Problem.get_problem().get_depot_by_id_depot(id_depot))
                
                if capacity_depot >= (request_cluster + request_customer):
                    request_cluster += request_customer
                    capacity_depot -= request_cluster
                    
                    cluster.set_request_cluster(request_cluster)
                    cluster.get_items_of_cluster().append(id_customer)
                    
                    self.calculate_attraction_coefficient(col, row, self.list_coefficients)
                    #self.scaled_matrix[row, col] = float('inf')
                    
                    self.list_customers_to_assign[col] = None
                    self.update_scaled_matrix(self.list_clusters, col, self.scaled_matrix, self.list_coefficients)
                    
                    if self.is_full_depot(self.list_customers_to_assign, request_cluster, capacity_depot):
                        self.scaled_matrix[pos_depot, :] = float('inf')
                        cost_matrix[pos_depot, :] = float('inf')
                        
                        for customer_id in self.list_clusters[pos_cluster].get_items_of_cluster():
                            pos_element_1 = Problem.get_problem().get_pos_element(customer_id)
                            self.scaled_matrix[:, pos_element_1] = float('inf')
                            cost_matrix[:, pos_element_1] = float('inf')
                            
                        if self.list_clusters[pos_cluster].get_items_of_cluster():
                            self.solution.get_clusters().append(self.list_clusters.pop(pos_cluster))
                        else:
                            self.list_clusters.pop(pos_cluster)
                    else:
                        if np.all(np.isinf(self.scaled_matrix[row, col])):
                            self.solution.get_unassigned_items().append(id_customer)
                            filtered_customers = [customer for customer in self.list_customers_to_assign if customer is not None]
                            self.list_customers_to_assign[Problem.get_problem().find_pos_customer(filtered_customers, id_customer)] = None
                        
                        self.scaled_matrix[:, col] = float('inf')
                        cost_matrix[:, col] = float('inf')
                        
    def finish(self):
        if self.list_customers_to_assign:
            for customer in self.list_customers_to_assign:
                if customer is not None:
                    self.solution.get_unassigned_items().append(customer.get_id_customer())

        if self.list_clusters:
            for cluster in self.list_clusters:
                if cluster.get_items_of_cluster():
                    self.solution.get_clusters().append(cluster)
        
        OSRMService.clear_distance_cache()
        
        return self.solution
    
    def calculate_attraction_coefficient(self, pos_customer: int, pos_element: int, coefficients: List[float]):
        current_att_coeff = 1.0
        new_att_coeff = 1.0
        
        if (pos_element + len(coefficients)) < len(coefficients):
            current_att_coeff = coefficients[pos_element]
            new_att_coeff = min(1, (current_att_coeff +(current_att_coeff * self.degradation_coefficient)))
        else:
            new_att_coeff = 1.0
        
        coefficients[pos_customer] = new_att_coeff
    
    def initialize_coefficients(self) -> List[float]:
        coefficients = [0.0] * Problem.get_problem().get_total_customers()
        return coefficients

    def fill_list_scaled_distances(self) -> List[List[float]]:
        scaled_distances = []
        cost_matrix: np.ndarray = self.initialize_cost_matrix(Problem.get_problem().get_customers(), Problem.get_problem().get_depots(), self.distance_type)
        
        list_id_customers = Problem.get_problem().get_list_id_customers()
        list_id_depots = Problem.get_problem().get_list_id_depots()
        
        id_to_index = {id_depot: idx for idx, id_depot in enumerate(list_id_depots)}
        indexed_depots = [id_to_index[id_depot] for id_depot in list_id_depots]
        
        total_items = len(list_id_customers)
        total_clusters = len(list_id_depots)
        
        for i in range(total_clusters):
            list_distances = [float('inf')] * (total_items)
            scaled_distances.append(list_distances)
        
        for i in range(total_clusters):
            pos_depot_matrix = indexed_depots[i]
                
            for j in range(total_items):
                pos_customer_matrix = Problem.get_problem().get_pos_element(list_id_customers[j])
                list_distances[j] = cost_matrix[pos_depot_matrix, pos_customer_matrix]
                
            #list_distances.extend([float('inf')] * total_clusters)
            scaled_distances[i] = list_distances
        
        return scaled_distances

    def initialize_scaled_matrix(self) -> np.ndarray:
        scaled_matrix: np.ndarray = np.full((4, 50), float('inf'))
        
        for i in range(len(self.list_scaled_distances)):
            for j in range(len(self.list_scaled_distances[i])):
                scaled_matrix[i, j] = self.list_scaled_distances[i][j]
        
        return scaled_matrix
    
    def update_scaled_matrix(self, clusters: List[Cluster], pos_customer: int, scaled_matrix: np.ndarray, coefficients: List[float]):
        cost_matrix: np.ndarray = self.initialize_cost_matrix(Problem.get_problem().get_customers(), Problem.get_problem().get_depots(), self.distance_type)
        
        for cluster in clusters:
            pos_cluster = (Problem.get_problem().get_pos_element(cluster.get_id_cluster()) - len(self.list_customers_to_assign))
            distance = cost_matrix[pos_cluster, pos_customer]
            scaled_distance = distance * coefficients[pos_customer]
            
            scaled_matrix[pos_cluster, pos_customer] = scaled_distance

    def get_pos_cluster(self, pos_customer: int, clusters: List[Cluster]):
        id_customer = Problem.get_problem().get_list_id_customers()[pos_customer]
        i = 0
        found: bool = False
        
        while i < len(clusters):
            j = 0
            if clusters[i].get_items_of_cluster():
                while j < len(clusters[i].get_items_of_cluster()) and not found:
                    if clusters[i].get_items_of_cluster()[j] == id_customer:
                        pos_cluster = i
                        found = True
                    else:
                        j += 1
            i += 1
        return pos_cluster